apps/yggdrasil/lib/yggdrasil/user.ex

```
@@ -29,14 +37,120 @@ defmodule Yggdrasil.User do
+
+  * Role1
+    * foo
+      * :read
+      * :write
+  * Role2
+    * foo
+      * :read
+      * :all
+
+  The set of permissions checked for foo will be `[:read, :write, :all]`
+
+  ## Examples
+      # checking a single resource
+      User.is_granted? user, foo: [:read]
+
+      # checking multiple resources is allowed as well
+      User.is_granted? user, foo: [:read], bar: [:read, :write]
   """
-  def hash_password(changeset = %{:valid? => false}) do
+  def is_granted!(user, res_perms) do
+    case is_granted?(user, res_perms) do
+      {:ok, result}  -> result
+      {:error, msgs} -> raise RuntimeError, inspect(msgs)
+    end
+  end
+
+  def is_granted?(_user, []),  do: build_error ["resource keyword list is empty"]
+  def is_granted?(_user, nil), do: build_error ["resource keyword list is nil"]
+  def is_granted?(user, res_perms) do
+    res_perms
+    |> Enum.map(fn {res, perms} -> {res, user.permissions[res], perms} end)
+    |> Enum.map(&has_permissions/1)
+    |> Enum.reduce({:ok, true}, fn
+      {:ok, _},  acc={:ok, false}   -> acc
+      {:ok, res},    {:ok, true}    -> {:ok, res}
+      {:ok, _},  acc={:error, _}    -> acc
+      {:error, msg}, {:ok, _}       -> {:error, [msg]}
+      {:error, msg}, {:error, msgs} -> {:error, [msg|msgs]}
+    end)
+  end
+
+  defp has_permissions({_res, nil, _requried}), do: {:ok, false}
+  defp has_permissions({res, _given, nil}),     do: build_error "required permission list is nil for #{res}"
+  defp has_permissions({res, [], _required}),   do: build_error "given permission list is empty for #{res}"
+  defp has_permissions({res, _given, []}),      do: build_error "required permission list is empty for #{res}"
+  defp has_permissions({_res, given, required}) do
+    given = MapSet.new given
+    required = MapSet.new required
+
+    {:ok, MapSet.subset?(required, given)}
+  end
+
+  defp build_error(msg) do
+    Logger.error msg
+    {:error, msg}
+  end
```

I still feel like this should just be is_granted?/2 function that raises an error

I don't think there is anything the caller could do or any recovering action. need to dig into ectos repo but I think for example even if you are using like Repo.all or the others that if there is db connection issue that it raise an exception instead of returning an error tuple. Think I witnessed it yesterday actually. 

I think error tuple are perfect for things that could fail when called normally, like trying to read a file or insert a record where handling the error for reading a file you may retry x number of times and with insert maybe it fails because of a unique constraint which the user will be able to correct by choosing a unique name. 

With is_granted?/2 you don't have that pattern really. If it raises an error it was called wrong/configured wrong and truly is a bad state that should halt and I think that is ok and idiomatic. Not everything uses error tuples. 

Don't know happy to leave as is for now, but I think I was on the right track with no error tuples and simply returned true or false and raised errors in those bad states.